coding.go

```diff
@@ -123,6 +140,11 @@ func (n *ProtoNode) GetPBNode() *pb.PBNode {
 		}
 	}
 
+	// Ensure links are sorted prior to encode, regardless of `linksDirty`. They
+	// may not have come sorted if we deserialized a badly encoded form that
+	// didn't have links already sorted.
+	sort.Stable(pbLinkSlice(pbn.Links))
```

Comment:
Wouldn't it be simpler if `Unmarshal` set the dirty bit so we would only resort the first time after unmarshal not always ?

Maybe we can't because unmarshaling is handled by other ipld code.
It's probably fine to do this go1.19's PDQSort will do a fast `O(N)` linear sweep if the data is already sorted.

Reply:
Yeah, this is tricky. The main point of this PR is to allow stable inspection of the deserialized form. If we set `linksDirty` on an unmarshal, then we're going to trigger a re-sort on a number of inspection operations, losing access to the underlying order of the links as they appear in the bytes.

But on top of that, we have to decide what to do on some non-passive inspection operations, including this one (`GetPBNode()`) and `Copy()`. I've decided that in that case, we should present a _properly_ sorted form, making it not quite a 1:1 "copy" of the data, but a "copy into a correct form".

Thankfully, there are increasingly limited uses of either `GetPBNode()` or `Copy()`, and even these entire call paths, due to replacement by go-ipld-prime and go-codec-dagpb.

But, there's one path where the nuance gets a bit tricker: `marshalImmutable()` gets called by both `Marshal()` and `EncodeProtobuf()`, for the former it just wants a `[]byte` form of the Node so it just wants the links sorted. That gets done on its way out by go-codec-dagpb regardless of what it's given, so that's fine. But for `EncodeProtobuf()` it's also caching a copy of the go-ipld-prime `Node` form that's built by `marshalImmutable()`, and then that can be used to pass this `ProtoNode` off as if it's a go-ipld-prime `Node`, when `Links` field in whatever order it was given.

I've revisited my original code for that last case and have removed the sorting from  `marshalImmutable()` and put it only on `EncodeProtobuf()` only if `linksDirty` is set. That way we're fixing sorting on mutation, but capturing the Link order as they are in the bytes even for the `Node` form of this `ProtoNode`.